### PR TITLE
AK: Add OOM safe interface to HashTable/Map

### PR DESCRIPTION
**Pull request by Hendiadyoin1**
Work towards OOM safety in the Kernel.

This adds a new HashSetResult only returned by try_set, to signal
allocation failure during setting.

### CODE REVIEW

File: AK/HashTable.h [hunk 1]
```diff
@@ -184,12 +185,16 @@ class HashTable {
...
     {
         for (size_t i = 0; i < N; ++i) {
-            set(from_array[i]);
+            if (try_set(from_array[i]) == HashSetResult::Failed)
+                return false;
         }
+        return true;
     }
+    template<typename U, size_t N>
+    void set_from(U (&from_array)[N]) { VERIFY(try_set_from(from_array)); }
```
Comment (Dexesttp) [inline AK/HashTable.h:197]: Please don't put side-effects into `VERIFY()` calls, as the calls could get removed altogether.
```suggestion
    void set_from(U (&from_array)[N])
    {
        bool result = try_set_from(from_array);
        VERIFY(result);
    }
```

File: AK/HashTable.h [hunk 2]
```diff
@@ -378,32 +392,33 @@ class HashTable {
...
 
         for (auto it = move(old_iter); it != end(); ++it) {
             insert_during_rehash(move(*it));
             it->~T();
         }
 
         kfree_sized(old_buckets, size_in_bytes(old_capacity));
+        return true;
     }
+    void rehash(size_t new_capacity) { VERIFY(try_rehash(new_capacity)); }
```
Comment (Dexesttp) [inline AK/HashTable.h:421]: Same here:
```suggestion
    void rehash(size_t new_capacity)
    {
        bool result = try_rehash(new_capacity);
        VERIFY(result);
    }
```

**Review by Dexesttp — COMMENTED** [review]
Usually, adding an API that nobody uses is a bad thing.

Here, it "kind of " makes sense since we want to move towards OOM handling in the Kernel. But maybe there's other solutions (like a non-allocating hashmap/hashtable, ala IntrusiveLinkedList) that would make this change not useful? I don't know what options are possible.

Anyways, I can't comment one way or another about that and I'll let somebody else, but here are some reviews about the code:

**Comment by bgianfo** [comment]
> Usually, adding an API that nobody uses is a bad thing.
> 
> Here, it "kind of " makes sense since we want to move towards OOM handling in the Kernel. 

I will be putting these to good use in the kernel the moment this is merged :+1:

**Comment by bgianfo** [comment]
Linking #6369 as this fixes an issue explicitly called out there.

File: AK/HashTable.h [hunk 3]
```diff
@@ -15,6 +15,7 @@
 namespace AK {
 
 enum class HashSetResult {
+    Failed = 0,
```
Comment (sin-ack) [inline AK/HashTable.h:18]: `Failed` doesn't really say anything. Did we fail to allocate, or did something else fail? Right now, we don't return a result for any other failure, but I feel like this could be more descriptive nevertheless.

File: AK/HashTable.h [hunk 4]
```diff
@@ -360,7 +378,7 @@ class HashTable {
         }
     }
 
-    [[nodiscard]] static size_t size_in_bytes(size_t capacity)
+    [[nodiscard]] static constexpr size_t size_in_bytes(size_t capacity)
```
Comment (sin-ack) [inline AK/HashTable.h:381]: This is a separate change, please move it to another commit.

File: AK/HashTable.h [hunk 5]
```diff
@@ -378,31 +396,36 @@ class HashTable {
         auto old_capacity = m_capacity;
         Iterator old_iter = begin();
 
-        if constexpr (IsOrdered) {
-            m_buckets = (BucketType*)kmalloc(size_in_bytes(new_capacity));
-            __builtin_memset(m_buckets, 0, size_in_bytes(new_capacity));
+        auto new_buckets = kmalloc(size_in_bytes(new_capacity));
+        if (!new_buckets)
+            return false;
```
Comment (sin-ack) [inline AK/HashTable.h:401]: The change to how `m_collection_data` is assigned is a change that shouldn't have been bundled into this commit, as it's hard to verify the diff now. Please move this to another commit.

**Review by sin-ack — CHANGES_REQUESTED** [review]

File: AK/HashTable.h (hunk 5 above)
Comment (Hendiadyoin1) [reply AK/HashTable.h:401]: This does not change how it is assigned, but only change the duplicate allocation lines into only one that honors allocation failure.
The other line was just moved down to cut down in if statements

File: AK/HashTable.h (hunk 3 above)
Comment (Hendiadyoin1) [reply AK/HashTable.h:18]: Setting a HashTable can only fail one way as far as I know

The only other thing I can think of are collisions and we just rehash in these cases

**Review by sin-ack — APPROVED** [review]
LGTM overall, the `m_collection_data` part still feels kinda wrong but that's more of a personal preference than anything.

**Comment by bgianfo** [comment]
Thanks for working on this! 🚀
